src/nvim/charset.c

```diff
@@ -461,8 +451,8 @@ char_u* str_foldcase(char_u *str, int orglen, char_u *buf, int buflen)
 
             if (olen != nlen) {
               if (buf == NULL) {
-                STRMOVE(GA_PTR(i) + nlen, GA_PTR(i) + olen);
-                ga.ga_len += nlen - olen;
+                STRMOVE(ks.s + i + nlen, ks.s + i + olen);
+                ks.l += nlen - olen;
```

Comment:
Looks like the existing garray macros and functions could be internally implemented as a wrapper around kstring. Then we wouldn't need to change any of the existing garray code, and could benchmark that.